platform/api.scripting/src/org/netbeans/modules/scripting/GraalEngine.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.scripting;
+
+import java.io.Reader;
+import javax.script.Bindings;
+import javax.script.Invocable;
+import javax.script.ScriptContext;
+import javax.script.ScriptEngine;
+import javax.script.ScriptException;
+import org.graalvm.polyglot.PolyglotException;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+
+final class GraalEngine implements ScriptEngine, Invocable {
+
+    private final GraalEngineFactory factory;
+
+    GraalEngine(GraalEngineFactory f) {
+        this.factory = f;
+    }
+
+    private String id() {
+        return factory.id;
+    }
+
+    @Override
+    public GraalContext getContext() {
+        return factory.ctx;
+    }
+
+    @Override
+    public GraalEngineFactory getFactory() {
+        return factory;
+    }
+
+    @Override
+    public Object eval(String src, ScriptContext arg1) throws ScriptException {
+        Value result = evalImpl(arg1, src);
+        return unbox(result);
+    }
+
+    private Value evalImpl(ScriptContext arg1, String src) throws ScriptException {
+
+        try {
+            return ((GraalContext) arg1).ctx().eval(id(), src);
+        } catch (PolyglotException e) {
+            throw new ScriptException(e);
+        }
+    }
+
+    @Override
+    public Object eval(Reader arg0, ScriptContext arg1) throws ScriptException {
+        Source src = Source.newBuilder(id(), arg0, null).buildLiteral();
+        Value result = ((GraalContext)arg1).ctx().eval(src);
+        return unbox(result);
+    }
+
+    @Override
+    public Object eval(String arg0) throws ScriptException {
+        return eval(arg0, factory.ctx);
+    }
+
+    @Override
+    public Object eval(Reader arg0) throws ScriptException {
+        return eval(arg0, factory.ctx);
+    }
+
+    @Override
+    public Object eval(String arg0, Bindings arg1) throws ScriptException {
+        return eval(arg0);
```

Comment:
maybe stupid idea, but couldn't be `ctx().getBindings(id()).putMember()` be used to place binding values into Engine scope ?
Otherwise I'd suggest to log a warning if `!arg1.isEmpty()`.

Reply:
It is not easy to support alternative bindings. Let's throw an exception: 26fbc8060f100